nano/core_test/node.cpp

```diff
@@ -2724,41 +2724,60 @@ TEST (node, fork_invalid_block_signature)
-	auto send2_corrupt (std::make_shared<nano::send_block> (*send2));
-	send2_corrupt->signature = nano::signature (123);
-	system.nodes[0]->process_active (send1);
+	nano::block_builder builder;
+	std::shared_ptr<nano::block> send1 = builder.state()
+				.account (nano::test_genesis_key.pub)
+				.previous (genesis.hash ())
+				.representative (nano::test_genesis_key.pub)
+				.balance (nano::genesis_amount - nano::Gxrb_ratio)
+				.link (nano::test_genesis_key.pub)
+				.work (*system.work.generate (genesis.hash ()))
+				.sign (nano::test_genesis_key.prv, nano::test_genesis_key.pub)
+				.build ();
+	std::shared_ptr<nano::block> send2 = builder.state()
+				.account (nano::test_genesis_key.pub)
+				.previous (genesis.hash ())
+				.representative (nano::test_genesis_key.pub)
+				.balance (nano::genesis_amount - 2 * nano::Gxrb_ratio)
+				.link (nano::test_genesis_key.pub)
+				.work (*system.work.generate (genesis.hash ()))
+				.sign (nano::test_genesis_key.prv, nano::test_genesis_key.pub)
+				.build ();
+	std::shared_ptr<nano::block> send3 = builder.state()
+				.account (nano::test_genesis_key.pub)
+				.previous (genesis.hash ())
+				.representative (nano::test_genesis_key.pub)
+				.balance (nano::genesis_amount - 2 * nano::Gxrb_ratio)
+				.link (nano::test_genesis_key.pub)
+				.work (*system.work.generate (genesis.hash ()))
+				.sign (nano::test_genesis_key.prv, 0) // Invalid signature
+				.build ();
+	auto channel1 (system.nodes[0]->network.udp_channels.create (system.nodes[0]->network.endpoint ()));
+	system.nodes[0]->network.process_message (nano::publish (send1), channel1);
 	system.deadline_set (5s);
-	while (!system.nodes[0]->block (send1->hash ()))
-	{
-		ASSERT_NO_ERROR (system.poll ());
-	}
-	system.nodes[0]->active.publish (send2_corrupt);
-	ASSERT_NO_ERROR (system.poll ());
-	system.nodes[0]->active.publish (send2);
-	std::vector<nano::block_hash> vote_blocks;
-	vote_blocks.push_back (send2->hash ());
-	auto vote (std::make_shared<nano::vote> (nano::test_genesis_key.pub, nano::test_genesis_key.prv, 0, vote_blocks));
-	{
-		auto transaction (system.nodes[0]->store.tx_begin_read ());
-		nano::unique_lock<std::mutex> lock (system.nodes[0]->active.mutex);
-		system.nodes[0]->vote_processor.vote_blocking (transaction, vote, std::make_shared<nano::transport::channel_udp> (system.nodes[0]->network.udp_channels, system.nodes[0]->network.endpoint (), system.nodes[0]->network_params.protocol.protocol_version));
-	}
-	while (system.nodes[0]->block (send1->hash ()))
+	std::shared_ptr<nano::election> election;
+	while (election == nullptr)
 	{
 		ASSERT_NO_ERROR (system.poll ());
+		nano::lock_guard<std::mutex> lock (system.nodes [0]->active.mutex);
+		auto existing = system.nodes [0]->active.blocks.find (send1->hash ());
+		if (existing != system.nodes [0]->active.blocks.end ())
+		{
+			election = existing->second;
+		}
 	}
-	while (!system.nodes[0]->block (send2->hash ()))
+	ASSERT_EQ (1, election->blocks.size ());
```

Comment:
Do we need active mutex lock to prevent TSAN issues here and later for election access?

Reply:
Fixed.